sphinx/source/docs/dev_guide/testing.rst

```diff
@@ -1,231 +1,447 @@
-* ``selenium``: a test requiring selenium
+    .. code-block:: sh
 
-Code Coverage
-~~~~~~~~~~~~~
+        pytest tests/integration
 
-To run any of the tests with coverage, use the following:
+    These tests mostly simulate UI interactions. Therefore, they require
+    `Chrome`_ or `Chromium`_ and `Selenium`_ with the `ChromeDriver`_ web
+    driver.
 
-.. code-block:: sh
+Code coverage
+    To create a coverage report, use ``pytest`` with the command line options
+    ``--cov=bokeh`` and ``--cov-config=tests/.coveragerc``:
 
-  pytest --cov=bokeh
+    .. code-block:: sh
 
-To report on a subset of the Bokeh package, pass e.g. ``-cov=bokeh/models``.
+        pytest --cov=bokeh --cov-config=tests/.coveragerc
 
-Other Options
-~~~~~~~~~~~~~
+    You also have the option to add
+    ``--cov=bokeh --cov-config=tests/.coveragerc`` when running a specific
+    subset of tests. This adds a coverage report to the test results. For
+    example:
 
-To run any of the tests without standard output captured use:
+    .. code-block:: sh
 
-.. code-block:: sh
+        pytest --cov=bokeh --cov-config=tests/.coveragerc tests/codebase
 
-  pytest -s
+    .. seealso::
+        Coverage reports use the pytest plugin `pytest-cov`_. For more
+        information, see the `documentation for pytest-cov`_.
 
-See the `pytest`_ documentation for further information on ``pytest`` and
-its options.
+Examples tests
+    The ``examples`` tests run a selection of the examples in the Bokeh
+    repository and generate images to compare against images generated by
+    previous releases.
 
-Examples Tests
-~~~~~~~~~~~~~~
+    The example tests use a specialized testing framework, including a custom
+    configuration of Chrome. Therefore, it is **recommended to not run those
+    tests locally**. Instead, :ref:`Bokeh's CI <devguide_testing_ci>` runs all
+    examples tests once you :ref:`create a Pull Request <devguide_pull_requests>`.
 
-The ``examples`` tests run a selection of the Bokeh examples and generate
-images to compare against previous releases. A report is generated that
-displays the current and previous images, as well as any image difference.
+    Running these tests also generates a report that displays
+    the current and previous images, as well as any image difference.
 
-.. note::
-    The tests do not currently fail if the images are different, the test
-    report must be inspected manually.
+    .. note::
+        The tests currently don't fail if the images are different. You need to
+        inspect the test report manually.
+
+    To run the examples tests, you first need to start a customized headless
+    version of Chrome in the background. This headless browser needs to be
+    started from the ``bokehjs`` folder. Use the following commands from the top
+    level of your *source checkout* directory:
+
+    .. code-block:: sh
+
+        cd bokehjs
+        node make test:spawn:headless
```

Comment:
@mattpap While this works fine on Ubuntu (20.04), I can't get the examples tests to pass on win10. Spawning the headless browser seems to look fine:

![image](https://user-images.githubusercontent.com/39711796/128809883-0ec6e943-a412-4f7b-92a6-c78c13bd3fb9.png)

However, while `pytest tests/test_examples.py` works fine on Ubuntu (Chromium 92.0.4515.131), 392 of the tests fail on win10 (92.0.4515.131) with this error:

```
Cannot connect to browser:  Error: connect ECONNREFUSED 127.0.0.1:9222
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1142:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 9222
}
```
This could well be some quirk in my local windows system - I was just wondering if there is something I should be doing differently on Windows than on Linux.

Reply:
`test:spawn:headless` perhaps needs some tweaking on windows. A possibly safer alternative is to use `node make test:run:headless`, just then you need two terminals open to proceed.